src/include/btree_cmp_inline.h

```diff
@@ -166,6 +302,129 @@ __wt_compare_bounds(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_ITEM *key, u
     return (0);
 }
 
+#ifdef HAVE_X86INTRIN_H
+/*
+ * __lex_compare_skip_gt_16 --
+ *     Lexicographic comparison routine for data greater than 16 bytes, skipping leading bytes.
+ *     Returns: < 0 if user_item is lexicographically < tree_item = 0 if user_item is
+ *     lexicographically = tree_item > 0 if user_item is lexicographically > tree_item We use the
+ *     names "user" and "tree" so it's clear in the btree code which the application is looking at
+ *     when we call its comparison function.
+ */
+static WT_INLINE int
+__lex_compare_skip_gt_16(
+  const uint8_t *ustartp, const uint8_t *tstartp, size_t len, int lencmp, size_t *matchp)
+{
+    __m128i res_eq, t, u;
+    int32_t eq_bits;
+    size_t match, final_bytes, final_match;
+
+    WT_ASSERT(NULL, len > WT_VECTOR_SIZE);
+
+    match = *matchp;
+    final_bytes = len - WT_VECTOR_SIZE;
+
+    /*
+     * Compare 16 bytes at a time until we find a difference or run out of 16 byte chunks to
+     * compare.
+     */
+    for (; match < final_bytes; match += WT_VECTOR_SIZE) {
+        u = _mm_loadu_si128((const __m128i *)(ustartp + match));
+        t = _mm_loadu_si128((const __m128i *)(tstartp + match));
+        res_eq = _mm_cmpeq_epi8(u, t);
+        if ((eq_bits = _mm_movemask_epi8(res_eq)) != WT_VECTOR_MASK)
+            goto final128;
+    }
+
+    /*
+     * Rewind until there is exactly 16 bytes left. We know we started with at least 16, so we are
+     * still in bound.
+     */
+    match = final_bytes;
+    u = _mm_loadu_si128((const __m128i *)(ustartp + final_bytes));
+    t = _mm_loadu_si128((const __m128i *)(tstartp + final_bytes));
+    res_eq = _mm_cmpeq_epi8(u, t);
+    eq_bits = _mm_movemask_epi8(res_eq);
+
+    if (eq_bits == WT_VECTOR_MASK) {
+        *matchp = len;
+        return (lencmp);
+    } else {
+final128:
+        /* The matching bits corresponds to the trailing bits in the result. */
+        final_match = (size_t)__builtin_ctz(~(uint32_t)eq_bits);
```

Comment:
Ugh. Why is there a `WT_LEADING_ZEROS` but no `WT_TRAILING_ZEROS`? You may need to make one for the windows build since `__builtin_ctz` doesn't exist there. I know the `_mm_` functions are the same on both msvc and clang/gcc since they are standardized by intel.

That said, it is super weird that WT_LEADING_ZEROS counts zero *bytes* rather than zero *bits*. That happens to be correct for the non-x86 version below, but this version needs bits. Maybe it is best to just make a local fix here rather than trying to clean up the universe today:

```suggestion
#ifndef _MSC_VER
        final_match = (size_t)__builtin_ctz(~(uint32_t)eq_bits);
#else
        {
            unsigned long res;
            _BitScanForward(&res, ~(uint32_t)eq_bits);
            final_match = res;
        }
#endif
```


PS, wouldn't it be great if you could just use C++: https://en.cppreference.com/w/cpp/numeric/countr_one 😁

Reply:
I'm kind of curious why the windows tests still work without this?